pkg/controller/networkpolicy/validate.go

```diff
@@ -694,3 +772,29 @@ func (g *groupValidator) updateValidate(curObj, oldObj interface{}, userInfo aut
 func (g *groupValidator) deleteValidate(oldObj interface{}, userInfo authenticationv1.UserInfo) (string, bool) {
 	return "", true
 }
+
+// createValidate validates the CREATE events of Service resources.
+func (s *serviceValidator) createValidate(curObj interface{}, userInfo authenticationv1.UserInfo) (string, bool) {
+	return "", true
+}
+
+// updateValidate validates the UPDATE events of Service resources.
+func (s *serviceValidator) updateValidate(curObj, oldObj interface{}, userInfo authenticationv1.UserInfo) (string, bool) {
+	return "", true
+}
+
+// deleteValidate validates the DELETE events of Service resources.
+func (s *serviceValidator) deleteValidate(oldObj interface{}, userInfo authenticationv1.UserInfo) (string, bool) {
+	oldSvc := oldObj.(*v1.Service)
+	svcNamespacedName := k8s.NamespacedName(oldSvc.Namespace, oldSvc.Name)
+	// Service referred in ToServices field of existing ACNPs/ANPs cannot be deleted.
```

Comment:
We removed the restriction that a ClusterGroup must exist before it can be used as a child group and the restriction that a ClusterGroup must exist before it can be used in an Antrea ClusterNetworkPolicy, so I wonder if the reference of service should be weak too. @abhiraut @jianjuns

Reply:
The reason I choose to use the strong restriction here is to avoid the race condition on `ServiceToGroupID` map. For example, if we create a policy before adding a Service, then when this Service is added, we need to reprocess the policy and update the `ServiceToGroupID` map. The proxier needs to update the map and at the same time reconciler needs to use the value in the map to reconcile the rule.

Reply:
/cc @antoninbas

Reply:
I think we should indeed use a weak reference as suggested by @tnqn. I believe we can find a way to implement it that way without too much complexity. I think Quan may have more / better ideas,  but you could requeue if the groupID is not available yet, or start a goroutine that queues Services when their groupID is set.

Reply:
> you could requeue if the groupID is not available yet

If the reason why groupID is no available is that this Service doesn't exist instead of the race condition issue, will the requeue solution cause this rule to be continually requeued?

> start a goroutine that queues Services when their groupID is set

Sorry I'm a little bit confused about this solution, could you explain it more. Thanks!

Based on your ideas, I have also come up with a solution. Could we use `wait.Poll` to see if groupID exist when reconciling the rule:
```
if len(rule.To.ToServices) > 0 {
	var addresses []types.Address
	for _, service := range rule.To.ToServices {
                 // Use wait.Poll to avoid race condition
		err := wait.Poll(100*time.Millisecond, 1*time.Second, func() (bool, error) {
			svcToGroup := r.ofClient.GetServiceToGroup()
			_, ok := svcToGroup[k8s.NamespacedName(service.Namespace, service.Name)]
			return ok, nil
		})
		if err != nil {
			svcToGroup := r.ofClient.GetServiceToGroup()
			if groupID, ok := svcToGroup[k8s.NamespacedName(service.Namespace, service.Name)]; ok {
				addresses = append(addresses, openflow.NewServiceGroupIDAddress(groupID))
			}
		}
	}
	ofRule.To = append(ofRule.To, addresses...)
}
```

Reply:
I just realized that my above thought couldn't solve the issue on the DELETE Service case.
Glad to hear your `goroutine queue Service` idea @antoninbas  and more/better ideas from @tnqn.
Thank you!

Reply:
I was thinking of adding a channel so that the NP controller can be notified every time a new groupID is assigned or released. When the controller is notified, it can take the necessary steps.

Reply:
Pushed a weak reference version using a channel to avoid race condition.

Reply:
The current solution looks good to me.